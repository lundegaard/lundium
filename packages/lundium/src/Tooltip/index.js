export { default } from './Tooltip';
export { TooltipPlacements, TooltipTriggers } from './Tooltip';
