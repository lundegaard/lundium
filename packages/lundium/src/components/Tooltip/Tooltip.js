import React, { Fragment, cloneElement, useEffect } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { usePopperTooltip } from 'react-popper-tooltip';
import { cx, noop } from 'ramda-extension';

import { useTheme } from '../ThemeProvider';

const getPlacement = (placement, isRTL) => {
	if (isRTL) {
		if (placement === TooltipPlacements.left) {
			return TooltipPlacements.right;
		}
		if (placement === TooltipPlacements.right) {
			return TooltipPlacements.left;
		}
	}
	return placement;
};

const canUseDOM = Boolean(
	typeof window !== 'undefined' && window.document && window.document.createElement
);

const TooltipTrigger = ({
	children,
	// Some defaults changed in the hook implementation.
	// For backward compatibility we have to override them here.
	closeOnReferenceHidden = true,
	defaultTooltipShown,
	getTriggerRef,
	modifiers,
	onVisibilityChange,
	placement = 'right',
	portalContainer = canUseDOM ? document.body : null,
	tooltip,
	tooltipShown,
	usePortal = canUseDOM,
	...rest
}) => {
	const {
		triggerRef,
		getArrowProps,
		getTooltipProps,
		setTooltipRef,
		setTriggerRef,
		visible,
		state,
	} = usePopperTooltip(
		{
			// Some props renamed in the hook implementation.
			defaultVisible: defaultTooltipShown,
			onVisibleChange: onVisibilityChange,
			visible: tooltipShown,
			closeOnTriggerHidden: closeOnReferenceHidden,
			...rest,
		},
		{
			placement,
			modifiers,
		}
	);
	const reference = children({
		// No longer required, for backward compatibility.
		getTriggerProps: props => props,
		triggerRef: setTriggerRef,
	});

	const popper = tooltip({
		tooltipRef: setTooltipRef,
		getArrowProps,
		getTooltipProps,
		placement: state ? state.placement : undefined,
	});

	useEffect(() => {
		if (typeof getTriggerRef === 'function') {
			getTriggerRef(triggerRef);
		}
	}, [triggerRef, getTriggerRef]);

	return (
		<Fragment>
			{reference}
			{visible ? (usePortal ? createPortal(popper, portalContainer) : popper) : null}
		</Fragment>
	);
};

export const TooltipPlacements = {
	top: 'top',
	right: 'right',
	bottom: 'bottom',
	left: 'left',
};

TooltipTrigger.propTypes = {
	children: PropTypes.node,
	closeOnReferenceHidden: PropTypes.bool,
	defaultTooltipShown: PropTypes.bool,
	getTriggerRef: PropTypes.bool,
	modifiers: PropTypes.array,
	onVisibilityChange: PropTypes.func,
	placement: PropTypes.oneOf(Object.keys(TooltipPlacements)),
	portalContainer: PropTypes.node,
	tooltip: PropTypes.node,
	tooltipShown: PropTypes.bool,
	usePortal: PropTypes.bool,
};

const Tooltip = ({
	isUpperCase = true,
	onTooltipVisibilityChange = noop,
	placement = 'bottom',
	showArrow = true,
	showTooltip = false,
	trigger = 'hover',
	children,
	tooltip,
	...rest
}) => {
	const { isRTL } = useTheme();
	const realPlacement = getPlacement(placement, isRTL);

	return (
		<TooltipTrigger
			{...rest}
			trigger={trigger}
			defaultTooltipShown={showTooltip}
			onVisibilityChange={onTooltipVisibilityChange}
			placement={realPlacement}
			tooltip={({ getArrowProps, arrowRef, getTooltipProps, tooltipRef }) => (
				<div
					{...getTooltipProps({
						ref: tooltipRef,
						className: cx('tooltip', {
							'upper-case': isUpperCase,
						}),
					})}
				>
					{showArrow && (
						<div
							{...getArrowProps({
								ref: arrowRef,
								className: 'tooltip-arrow',
								'data-placement': realPlacement,
							})}
						/>
					)}
					<div className="tooltip-body">{tooltip}</div>
				</div>
			)}
		>
			{({ getTriggerProps, triggerRef }) =>
				cloneElement(children, {
					ref: triggerRef,
					...getTriggerProps(),
				})
			}
		</TooltipTrigger>
	);
};

export const TooltipTriggers = ['hover', 'click', 'right-click'];

Tooltip.propTypes = {
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	/** Tooltip upper case or not */
	isUpperCase: PropTypes.bool,
	/** Function called on state change */
	onTooltipVisibilityChange: PropTypes.func,
	/** Tooltip position - top, right, bottom or left */
	placement: PropTypes.oneOf(Object.keys(TooltipPlacements)),
	/** Show tooltip arrow or not */
	showArrow: PropTypes.bool,
	/** Show tooltip by default or not */
	showTooltip: PropTypes.bool,
	/** Content of tooltip */
	tooltip: PropTypes.node,
	/** Tooltip trigger - hover, click or right-click */
	trigger: PropTypes.oneOf(TooltipTriggers),
};

export default Tooltip;
