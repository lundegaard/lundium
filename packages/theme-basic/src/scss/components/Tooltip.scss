@import '../core';

$bgColor: rgb(33, 37, 41);

.tooltip {
	background-color: $bgColor;
	opacity: .9;
	font-size: 12px;
	color: rgb(255, 255, 255);
	border-radius: 4px;
	box-shadow: 0 4px 10px rgba(0, 0, 0, .2);
	display: flex;
	flex-direction: column;
	justify-content: center;
	margin: 1.3rem .4rem 1rem;
	padding: .4rem 1rem;
	transition: opacity .3s;
	// Popover needs to be fixed to work with touch events
	z-index: $zindex-tooltip;

	&.upper-case {
		text-transform: uppercase;
	}
}

.tooltip-arrow {
	height: 1rem;
	position: absolute;
	width: 1rem;
	color: $bgColor;
}

.tooltip-arrow::before {
	border-style: solid;
	content: '';
	display: block;
	height: 0;
	margin: auto;
	width: 0;
}

.tooltip-arrow::after {
	border-style: solid;
	content: '';
	display: block;
	height: 0;
	margin: auto;
	position: absolute;
	width: 0;
}

.tooltip-arrow[data-placement*='bottom'] {
	height: 1rem;
	left: 0;
	margin-top: -.4rem;
	top: 0;
	width: 1rem;
}

.tooltip-arrow[data-placement*='bottom']::before {
	border-color: transparent transparent $bgColor transparent;
	border-width: 0 .5rem .4rem;
	position: absolute;
	top: -1px;
}

.tooltip-arrow[data-placement*='bottom']::after {
	border-color: transparent transparent $bgColor transparent;
	border-width: 0 .5rem .4rem;
}

.tooltip-arrow[data-placement*='top'] {
	bottom: 0;
	height: 1rem;
	left: 0;
	margin-bottom: -1rem;
	width: 1rem;
}

.tooltip-arrow[data-placement*='top']::before {
	border-color: $bgColor transparent transparent transparent;
	border-width: .4rem .5rem 0;
	position: absolute;
	top: 1px;
}

.tooltip-arrow[data-placement*='top']::after {
	border-color: $bgColor transparent transparent transparent;
	border-width: .4rem .5rem 0;
}

.tooltip-arrow[data-placement*='right'] {
	height: 1rem;
	left: 0;
	margin-left: -.7rem;
	width: 1rem;
}

.tooltip-arrow[data-placement*='right']::before {
	border-color: transparent $bgColor transparent transparent;
	border-width: .5rem .4rem .5rem 0;
}

.tooltip-arrow[data-placement*='right']::after {
	border-color: transparent $bgColor transparent transparent;
	border-width: .5rem .4rem .5rem 0;
	left: 6px;
	top: 0;
}

.tooltip-arrow[data-placement*='left'] {
	height: 1rem;
	margin-right: -.7rem;
	right: 0;
	width: 1rem;
}

.tooltip-arrow[data-placement*='left']::before {
	border-color: transparent transparent transparent $bgColor;
	border-width: .5rem 0 .5rem .4em;
}

.tooltip-arrow[data-placement*='left']::after {
	border-color: transparent transparent transparent $bgColor;
	border-width: .5rem 0 .5rem .4em;
	left: 4px;
	top: 0;
}
