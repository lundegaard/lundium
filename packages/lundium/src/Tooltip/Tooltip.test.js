import React from 'react';
import { shallow } from 'enzyme';

import Tooltip from './Tooltip';

describe('<Tooltip />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Tooltip {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
