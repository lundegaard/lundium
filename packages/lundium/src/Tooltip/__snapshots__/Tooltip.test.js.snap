// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> matches snapshot when shallowly rendered 1`] = `
<TooltipTrigger
  defaultTooltipShown={false}
  onVisibilityChange={[Function]}
  placement="bottom"
  tooltip={[Function]}
  trigger="hover"
>
  <Component />
</TooltipTrigger>
`;
